nengo_loihi/builder/builder.py

```diff
@@ -0,0 +1,177 @@
+        Mapping from objects to the integer seed assigned to that object.
+    """
+    def __init__(self, dt=0.001, label=None, builder=None):
+        self.dt = dt
+        self.label = label
+
+        self.inputs = collections.OrderedDict()
+        self.groups = collections.OrderedDict()
+
+        self.objs = collections.defaultdict(dict)
+        self.params = {}  # Holds data generated when building objects
+        self.probes = []
+        self.probe_conns = {}
+
+        self.seeds = {}
+        self.seeded = {}
+
+        self.builder = Builder() if builder is None else builder
+        self.build_callback = None
+
+        # --- other (typically standard) parameters
+        # Filter on decode neurons
+        self.decode_tau = 0.005
+        # ^TODO: how to choose this filter? Even though the input is spikes,
+        # it may not be absolutely necessary since tau_rc provides a filter,
+        # and maybe we don't want double filtering if connection has a filter
+
+        self.decode_neurons = Preset10DecodeNeurons(dt=dt)
+        self.node_neurons = OnOffDecodeNeurons(dt=dt)
+
+        # voltage threshold for non-spiking neurons (i.e. voltage decoders)
+        self.vth_nonspiking = 10
+
+        # limit for clipping intercepts, to avoid neurons with high gains
+        self.intercept_limit = 0.95
+
+        # scaling for PES errors, before rounding and clipping to -127..127
+        self.pes_error_scale = 100.
+
+        # learning weight exponent for PES (controls the maximum weight
+        # magnitude/weight resolution)
+        self.pes_wgt_exp = 4
+
+        # Will be provided by Simulator
+        self.chip2host_params = {}
+
+    def __getstate__(self):
+        raise NotImplementedError("Can't pickle nengo_loihi.builder.Model")
+
+    def __setstate__(self, state):
+        raise NotImplementedError("Can't pickle nengo_loihi.builder.Model")
+
+    def __str__(self):
+        return "Model: %s" % self.label
+
+    def add_input(self, input):
+        assert isinstance(input, SpikeInput)
+        assert input not in self.inputs
+        self.inputs[input] = len(self.inputs)
+
+    def add_group(self, group):
+        assert isinstance(group, NeuronGroup)
+        assert group not in self.groups
+        self.groups[group] = len(self.groups)
+
+    def build(self, obj, *args, **kwargs):
+        built = self.builder.build(self, obj, *args, **kwargs)
+        if self.build_callback is not None:
+            self.build_callback(obj)
+        return built
+
+    def discretize(self):
+        for group in self.groups:
+            group.discretize()
+
+    def has_built(self, obj):
+        return obj in self.params
+
+    def validate(self):
+        if len(self.groups) == 0:
+            raise BuildError("No neurons marked for execution on-chip. "
+                             "Please mark some ensembles as on-chip.")
+
+        for group in self.groups:
+            group.validate()
+
+
+class Builder(object):
```

Comment:
Since this essentially is the same as the Nengo builder but using our registered functions, in #30 I essentially cloned the class by making this a mostly empty subclass (see https://github.com/nengo/nengo-loihi/pull/30/commits/ff01e4da120919039dcbbb9ac9341d310bac1b08#diff-75a6663f57ab43eda3b594e818e01e0dR88). Did something change in here such that it's not a straight copy of the Nengo builder?